### DISCUSSION TITLE
Ignorecommand with Regex to ignore devices in the same dynamic IPv6 subnet

### DISCUSSION THREAD

**Discussion by Giresharu** [fail2ban/fail2ban#3899] [Show and tell]
After enabling IPv6, some devices will prefer to use IPv6 for connections, and this includes devices within the local area network. If both the connecting and the connected devices have public IPv6 addresses that are periodically changed by the operator, it is not possible to fill in a fixed subnet in `ignoreip` for ignoring.

At this point, we can use `ignorecommand` with regular expressions to obtain the corresponding subnet for dynamic ignoring. Here is the code I used:

```
ignorecommand = /usr/bin/ip -6 route show | grep -oP '^[0-9a-z:\/]+(?= dev enp1s0)'
# enp1s0 is my ethernet interface name, you should change it to yours.
```

Please note that it may not be universally applicable to all Linux devices and network situations, so if you cannot directly use the regular expression I provide, you can try to read the thought process and match the subnet by yourself.

First, I use 
```
ip -6 route show
```
This line of code can view the routing information of various network interfaces, including some virtual network interface (such as Docker, etc.), like this:
```
240e:xxxx:xxxx:xxxx::/64 dev enp1s0 proto ra metric 100 expires 259059sec mtu 1492 hoplimit 64 pref medium
fe80::/64 dev enp1s0 proto kernel metric 256 pref medium
fe80::/64 dev docker0 proto kernel metric 256 pref medium
fe80::/64 dev veth1549d38 proto kernel metric 256 pref medium
fe80::/64 dev veth20528f5 proto kernel metric 256 pref medium
fe80::/64 dev veth5c0a75f proto kernel metric 256 pref medium
fe80::/64 dev veth8f5f0f6 proto kernel metric 256 pref medium
default via fe80::xxxx:xxxx:xxxx:xxxx dev enp1s0 proto ra metric 100 expires 1753sec mtu 1492 hoplimit 64 pref medium
```
By observation, it is not difficult to find that the name of my physical network interface is called `enp1s0` (yours may be called `eth0` or something else), and it is located in a subnet starting with `240e` (public network of China Telecom), as well as the `fe80` local network subnet. We can match them with regular expressions.

First, we need to match content that has one thing in common, which is that they are all located before ` dev enp1s0`, so I first write down:
```
(?= dev enp1s0) 
# Don't forget there is a space before dev
```
This means that the content we want to match must be before ` dev enp1s0`.

Then we analyze the format of the subnet. First, they all start from the beginning of the line, contain numbers and letters, and include colons and slashes. Generally, there are no other things that contain the same elements but are not subnets, so we don't have to spend too much time describing the format, as long as they contain these elements:
```
^[0-9a-z:\/]+(?= dev enp1s0)
```
Where `^` indicates matching the beginning of the line, `[0-9a-z:\/]+` indicates matching one or more characters composed of `0-9`, `a-z`, as well as `:` and `/`, and then followed by the previously written `(?= dev enp1s0)` can correctly match the required subnet.

Let's test this regular expression in grep first:
```
ip -6 route show | grep -oP '^[0-9a-z:\/]+(?= dev enp1s0)'
```
The `-oP` options in grep stand for outputting the content matched by the regular expression.

On my system, this command yields the following results:
```
240e:xxxx:xxxx:xxxx::/64
fe80::/64
```

We just need to use this command for `ignorecommand` as well to ignore these two subnets.
However, I am not sure what environment variables it can access, so I chose to include the absolute path to the ip command as well:
```
ignorecommand = /usr/bin/ip -6 route show | grep -oP '^[0-9a-z:\/]+(?= dev enp1s0)'
```
Then restart fail2ban:
```
sudo fail2ban-client reload
```

Next, let's test with IPv6 devices in the same subnet!

**Comment by sebres** [comment]
1. I don't follow how you want to check current IP with that (since the tag `<ip>` or something else is unused in the command at all).
So in my opinion the command always doing the same for all IPs and so always exits with 0 (so would never ignore something).
I guess it is not intended... Please read the manual for [ignorecommand](https://manpages.debian.org/unstable/fail2ban/jail.conf.5.en.html#ignorecommand).
2. Also consider the cache to speedup lookup a bit, using [ignorecache](https://manpages.debian.org/unstable/fail2ban/jail.conf.5.en.html#ignorecache) parameter.
For example, see https://github.com/fail2ban/fail2ban/issues/2013#issuecomment-426940189 answering similar question.

**Reply by Giresharu** [reply]
Oh, I was negligent. I only noticed that the test "not banned" was successful, and I didn't find that this wrong writing caused all IPs to not be banned.🤣🤣
Just pretend I didn't share this, I'll find some time to look into it again.
